text/0000-incremental-compilation.md

```diff
@@ -0,0 +1,339 @@
+{
+    val.transmogrify()        
+}
+```
+
+In this example `transmogrify<Kid, Tiger>` will have a different dependency graph than `transmogrify<Dinosaur, Gastropod>`. In other words, the monomorphized implementation of `transmogrify<Kid, Tiger>` is not affected if the definition of `Dinosaur` or `Gastropod` changes and the dependency graph should reflect this.
+
+One way to model this behavior is by not creating dependency graph nodes for generic `program items` but `node templates`, which---like generic items---have type parameters and yield a concrete, monomorphic dependency graph node, if all type parameters are substituted for concrete arguments. When the need arises to check if a particular monomorphized function implementation from the cache can be re-used, the dependency graph for the function can be constructed on demand from the given `node template` and parameter substitutions.
+
+For consistency it also makes sense to treat non-generic `program items` as generic `program items` that happen to have zero type parameters. We thus obtain a direct correspondence between `program items` and `node templates` on the one hand and monomorphized `program item` instances and dependency graph nodes on the other.
+
+## Program Item and Compilation Artifact Identifiers
+`Program items` and their corresponding `node templates` must be identifiable in a way that is stable across multiple invocations of the compiler. The `DefId` type that is currently used for cross-crate item identification does not fit this requirement unfortunately, since it contains the rather unstable AST `NodeId`. Adding a single AST node currently can invalidate all `NodeId`s in the codebase due to the sequential Node ID assignment strategy.
+
+One straightforward way of creating stable identifiers would be to use the `program item`'s path within the AST, while generating local integer IDs for anonymous blocks:
+
+```rust
+// ID: "M1"
+mod M1 { 
+
+    // ID: "M1::f1"
+    fn f1() {
+
+        // ID: "M1::f1::S1"
+        struct S1;
+
+        {
+            // ID: "M1::f1::0::S2"
+            struct S2;
+
+        }
+
+        {
+            // ID: "M1::f1::1::E1"
+            enum E1;
+        }
+    }
+
+    impl SomeTrait for SomeType {
+        fn something() { // ID <SomeType as M1::SomeTrait>::something
+            ...
+        }
+    }
+}
+```
+
+The dependency tracking system as described above contains `node templates` for `program item` definitions on a syntactic level, that is, for each `struct`, `enum`, `type`, `trait`, there is one `node template`, for each `fn`, `static`, and `const` there are two (one for the interface, one for the body). However, as seen in the section on generics, the codebase can refer to monomorphized instances of program items that cannot be identified by a single identifier as described above. A reference like `Option<String>` is a composite of multiple `program item` IDs, a tree of program item IDs in the general case:
```

Comment:
Yes, that's a problem. I'd probably try to find a more stable symbol naming scheme.